### PR TITLE
Feature/alphanum marker

### LINKED ISSUE #54: Add alphanum validation rule
### Description
Add an `alphanum` (alphanumeric) validation rule that ensures a string contains only alphabetic characters (a-z, A-Z) and numeric characters (0-9).

### Usage Example
```go
type Product struct {
    ProductCode string // +govalid:alphanum
    SerialNumber string // +govalid:alphanum
}
```

### Expected Behavior
- Valid: "ABC123", "product2024", "12345", "abcdef", "X1"
- Invalid: "ABC-123", "product_2024", "12.34", " ABC123", "ABC 123", ""

### Implementation Notes
- Should validate that the string contains only letters (a-z, A-Z) and digits (0-9)
- No spaces, punctuation, or special characters allowed
- Empty strings should be invalid
- Should support any combination of letters and numbers
- Similar to go-playground/validator's `alphanum` tag

### Rationale
This is commonly used for:
- Product codes and SKUs
- Username validation
- Reference numbers
- API keys and tokens (simplified format)
- Alphanumeric identifiers without special characters

### PR DESCRIPTION
**Pull request by taua-almeida**
## Description

This PR adds the `alphanum` validation marker to govalid, which ensures that string fields contain only alphanumeric characters (letters a-z, A-Z and numbers 0-9). The validator follows govalid's zero-allocation design philosophy and provides high-performance validation for use cases like usernames, product codes, and serial numbers.

## Type of Change

- [X] Documentation update
- [X] Refactoring (no functional changes)

## Quality Checklist

Before submitting this pull request, ensure your contribution meets these quality standards:

### Code Quality
- [X] Code passes `make golangci-lint` without errors
- [X] All tests pass: `make test`
- [X] Code follows existing patterns and conventions
- [X] No hard-coded values (use constants or configuration)
- [X] Error messages are clear and actionable

### Testing Requirements
- [X] Golden tests created and passing (`internal/analyzers/govalid/testdata/`)
- [X] Unit tests implemented with boundary value testing (`test/unit/`)
- [X] Benchmark tests comparing against popular validation libraries (`test/benchmark/`)
  - [X] [go-playground/validator](https://github.com/go-playground/validator) (BenchmarkGoPlayground*)
  - [X] [asaskevich/govalidator](https://github.com/asaskevich/govalidator) (BenchmarkGoValidator*)
  - [X] [go-ozzo/ozzo-validation](https://github.com/go-ozzo/ozzo-validation) (BenchmarkOzzoValidation*)
  - [X] [gookit/validate](https://github.com/gookit/validate) (BenchmarkGookitValidate*)
- [X] All test files generated: `cd test && go generate`

### Performance Standards
- [X] Zero allocations in validation logic (verified via benchmarks)
- [X] Performance improvement over existing validation libraries (minimum 2x faster)
- [X] Benchmark results added to `test/benchmark/README.md`

### Documentation
- [X] README.md updated with new marker documentation
- [X] Code includes appropriate comments and examples
- [X] Validator usage examples provided

### Integration
- [X] Validator scaffold generated with `make generate-validator MARKER=yourmarker`
- [X] Registry files automatically updated
- [X] Binary rebuilt and installed: `go install ./cmd/govalid/`

### Final Verification
- [ ] All GitHub Actions CI checks pass
- [X] No breaking changes to existing functionality
- [ ] Contribution follows project license requirements

## Additional Notes

The `alphanum` validator implements efficient character-by-character validation using a simple range check approach:

**Key Features:**
- Zero allocations: Uses character-by-character iteration with no string operations
- Empty string handling: Returns false for empty strings (ensures required-like behavior)
- Performance optimized: Simple range comparisons that compile to efficient machine code

**Validation Behavior:**
- Accepts: "abc123", "ABC123", "123abc", "a1b2c3"
- Rejects: "", "abc 123", "abc-123", "abc@123", "abc_123"

## Related Issues
Closes #54

### CODE REVIEW

File: analyzers/govalid/testdata/src/length/govalid.golden
Comment (ccoVeille) [inline analyzers/govalid/testdata/src/length/govalid.golden:1]: These files are unrelated to the PR.

It appears to be a cleanup, it might be legitimate, but They should be removed from this PR, and addressed in another PR

File: internal/validator/rules/alphanum.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
+// Code generated by generate-validators; DO NOT EDIT.
```
Comment (ccoVeille) [inline internal/validator/rules/alphanum.go:1]: This file is not generated, but the implementation.

I was struggling to find where was the code you added, while GitHub diff was hidding me generated files

Please check for other files in this PR too.

File: internal/validator/rules/alphanum.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	return strings.ReplaceAll("Err@AlphanumValidation", "@", v.structName+v.FieldName())
+}
+
+func (v *alphanumValidator) Imports() []string {
+	return []string{}
+}
+
+// ValidateAlphanum creates a new alphanum validator for the given field.
+func ValidateAlphanum(pass *codegen.Pass, field *ast.Field, expressions map[string]string, structName string) validator.Validator {
+	validator.GeneratorMemory[fmt.Sprintf(alphanumKey, structName+field.Names[0].Name)] = false
```
Comment (BlackBuck) [inline internal/validator/rules/alphanum.go:84]: You could use `v.FieldName()` here instead of `field.Names[0].Name` here... it'll improve readability

Comment (BlackBuck) [inline internal/validator/rules/alphanum.go:84]: You could use v.FieldName() here instead of field.Names[0].Name here... it'll improve readability

File: test/marker_length_validator.go
Comment (BlackBuck) [inline test/marker_length_validator.go:1]: Why change this file in this feature branch?

File: test/marker_alphanum_validator.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	}
+
+	// ErrAlphanumProductCodeAlphanumValidation is returned when the AlphanumProductCode fails alphanum validation.
+	ErrAlphanumProductCodeAlphanumValidation = errors.New("field AlphanumProductCode failed alphanum validation")
+
+	// ErrAlphanumSerialNumberAlphanumValidation is returned when the AlphanumSerialNumber fails alphanum validation.
+	ErrAlphanumSerialNumberAlphanumValidation = errors.New("field AlphanumSerialNumber failed alphanum validation")
+
+	// ErrAlphanumUsernameAlphanumValidation is returned when the AlphanumUsername fails alphanum validation.
+	ErrAlphanumUsernameAlphanumValidation = errors.New("field AlphanumUsername failed alphanum validation")
```
Comment (BlackBuck) [inline test/marker_alphanum_validator.go:30]: I've noticed that there's a significant difference in the generated code between the older validators and the newer ones. Should we keep the same format as the previous ones, or should we refactor the older ones according to the two new validators?

Comment (taua-almeida) [reply test/marker_alphanum_validator.go:30]: In theory this new one is follow the standards introduced [here](https://github.com/sivchari/govalid/pull/90).
So I _would_ suggest us updating the old ones as a new issue.
What are your thoughts @sivchari ?

File: internal/validator/rules/alphanum.go (hunk 1 above)
Comment (ccoVeille) [reply internal/validator/rules/alphanum.go:1]: Let me provide context

I opened a PR, and nothing was adding code in the diff. Almost everything was generated.

So I looked for the file that was adding code, and I found this one.

But then I wondered, before dropping my initial comment, whether how were the other files in the same directory and I checked

https://github.com/search?q=repo%3Asivchari%2Fgovalid%20path%3Ainternal%2Fvalidator%2Frules%20generated&type=code

And none were using generated code in this folder, this convinced me some the was wrong.

So now, you somehow ask me to check what I checked before opening my comment.. So I'm 🤔


I might be wrong, and it won't be a problem.

But because there is chance something is wrong, I prefer to insist.

And if I'm wrong, where is the file with non generated code header that adds a feature to this PR ? Thanks

File: analyzers/govalid/testdata/src/length/govalid.golden
Comment (sivchari) [reply analyzers/govalid/testdata/src/length/govalid.golden:1]: It's curious. I've chcked that length marker is merged correctly and generated codes including with length marker. So what I expect is that these codes aren't deleted.
Please let me know about how can I reproduce it. I'd like to create pr to follow up.

File: test/marker_alphanum_validator.go (hunk 3 above)
Comment (sivchari) [reply test/marker_alphanum_validator.go:30]: @BlackBuck 
Is what you indicated difference isAlphaNumeric function?
If so, I think it should be moved to validationhelper because if the two files which is located in same directory uses alphanum, the generated code will be conflicted since isAlphaNumeric is defined two times.

Comment (BlackBuck) [reply test/marker_alphanum_validator.go:30]: Yeah. Previously, all the helper functions were defined in the `validationhelper` package, so I figured this should go there as well. I also noticed that the errors are named strangely. The other day, when I was working on `alpha` marker after the rebase, I saw error named `ErrEmailEmailEmailValidator`. I'm a bit confused... is that expected behavior?

Comment (sivchari) [reply test/marker_alphanum_validator.go:30]: Please see [it](https://github.com/sivchari/govalid/pull/79#issuecomment-3125516135)

File: internal/validator/rules/alphanum.go (hunk 2 above)
Comment (taua-almeida) [reply internal/validator/rules/alphanum.go:84]: So, I was revisiting this after these two days, here we don't have `v *alphanumValidator` we are getting from the `ast` itself.

**Review by ccoVeille — APPROVED** [review]

**Comment by sivchari** [comment]
Please rebase it 🙏

File: test/benchmark/README.md
Comment (sivchari) [inline test/benchmark/README.md:1]: Please revert it, this documentation would be updated later/

**Comment by sivchari** [comment]
When you run `make generate-validator`, all conflicts other than README will be resolved.

File: test/benchmark/benchmark_alphanum_test.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	for b.Loop() {
+		err := validate.Struct(&instance)
+		if err != nil {
+			b.Fatal("unexpected error:", err)
+		}
+	}
+	b.StopTimer()
+}
+
+func BenchmarkOzzoValidationAlphanum(b *testing.B) {
```
Comment (sivchari) [inline test/benchmark/benchmark_alphanum_test.go:52]: Let's remove this benchmark function. We decide not to support it.

File: test/benchmark/benchmark_alphanum_test.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	instance := test.Alphanum{ProductCode: "product2024", SerialNumber: "ABC123", Username: "username"}
+	b.ResetTimer()
+	for b.Loop() {
+		err := test.ValidateAlphanum(&instance)
+		if err != nil {
+			b.Fatal("unexpected error:", err)
+		}
+	}
+	b.StopTimer()
+}
```
Comment (sivchari) [inline test/benchmark/benchmark_alphanum_test.go:24]:
```suggestion
func BenchmarkGoValidAlphanum(b *testing.B) {
	instance := test.Alphanum{ProductCode: "product2024", SerialNumber: "ABC123", Username: "username"}
	for b.Loop() {
		err := test.ValidateAlphanum(&instance)
		if err != nil {
			b.Fatal("unexpected error:", err)
		}
	}
}
```

b.ResetTimer and b.StopTimer isn't necessary anymore. So please remove it.

**Review by sivchari — COMMENTED** [review]
Left final nits. I'd merge it after the suggestion is reflected.

File: MARKERS.md [hunk 6]
```diff
@@ -382,6 +382,45 @@ govalid supports the following markers:
...
+  ```
+- **Generated Code**:
+  ```go
+  func ValidateUser(t *User) error {
+      if t == nil {
+          return ErrNilUser
+      }
+
+      if !validationhelper.IsValidAlphanum(t.Username) {
+          return ErrUsernameAlphanumValidation
```
Comment (BlackBuck) [inline MARKERS.md:408]:
```suggestion
          return ErrUserUsernameAlphanumValidation
```

The errors have struct name as well as the marker name in them.

File: MARKERS.md [hunk 7]
```diff
@@ -382,6 +382,45 @@ govalid supports the following markers:
...
+      }
+
+      if !validationhelper.IsValidAlphanum(t.ProductCode) {
+          return ErrProductCodeAlphanumValidation
```
Comment (BlackBuck) [inline MARKERS.md:412]:
```suggestion
          return ErrUserProductCodeAlphanumValidation
```

File: MARKERS.md [hunk 8]
```diff
@@ -382,6 +382,45 @@ govalid supports the following markers:
...
+      }
+
+      if !validationhelper.IsValidAlphanum(t.SerialNumber) {
+          return ErrSerialNumberAlphanumValidation
```
Comment (BlackBuck) [inline MARKERS.md:416]:
```suggestion
          return ErrUserSerialNumberAlphanumValidation
```

File: docs/content/validators.md
Comment (BlackBuck) [inline docs/content/validators.md:1]: Similarly, rename the errors here as well...

File: internal/markers/markers_generated.go [hunk 9]
```diff
@@ -50,27 +56,26 @@ var (
...
+	GoValidMarkerMinitems: {},
 	GoValidMarkerMinlength: {},
-	GoValidMarkerRequired:  {},
-	GoValidMarkerUrl:       {},
-	GoValidMarkerUuid:      {},
-	GoValidMarkerNumeric:   {},
+	GoValidMarkerNumeric: {},
+	GoValidMarkerRequired: {},
+	GoValidMarkerUrl: {},
+	GoValidMarkerUuid: {},
```
Comment (BlackBuck) [inline internal/markers/markers_generated.go:80]: The formattign is off, I believe

File: validation/validationhelper/alphanum.go [hunk 10]
```diff
@@ -0,0 +1,15 @@
+package validationhelper
+
+// IsValidAlphanum checks if the string contains only alphanumeric characters using regex.
```
Comment (BlackBuck) [inline validation/validationhelper/alphanum.go:3]: nit: no regex here :)

File: internal/validator/rules/alphanum.go [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+	key := fmt.Sprintf(alphanumKey, v.structName+fieldName)
+	if validator.GeneratorMemory[key] {
+		return result.String()
+	}
+
+	validator.GeneratorMemory[key] = true
+
+	result.WriteString(strings.ReplaceAll(`
+	// Err@AlphanumValidation is returned when the @ fails alphanum validation.
+	Err@AlphanumValidation = errors.New("field @ failed alphanum validation")`, "@", v.structName+fieldName))
```
Comment (BlackBuck) [inline internal/validator/rules/alphanum.go:48]: Returning errors like this has a caveat that I think needs to be addressed. It is causing confusion across multiple PRs.
The issue here is that were returning the following error string:

```go
	result.WriteString(strings.ReplaceAll(`
	// Err@AlphanumValidation is returned when the @ fails alphanum validation.
	Err@AlphanumValidation = errors.New("field @ failed alphanum validation")`, "@", v.structName+fieldName))
```

So, if the `structName` is, say, `User`, and the `fieldName` is `Username`, then this would return:

```go
	// ErrUserUsernameAlphanumValidation is returned when the UserUsername fails alphanum validation.
	ErrUserUsernameAlphanumValidation = errors.New("field UserUsername failed alphanum validation")
```

The `UserUsername` part is causing confusion. I suggest having a `strings.Replacer` instead of `strings.Builder`. It would help us have something like "field User.Username failed alphanum validation" in a much cleaner way.

We could have a follow-up PR for it instead of changing it here if everyone agrees. 

WDYT @sivchari @ccoVeille @taua-almeida?

**Review by BlackBuck — CHANGES_REQUESTED** [review]
A few minor findings. Otherwise, good to go.

**Comment by sivchari** [comment]
Hi, @taua-almeida ! If you have time, please rebase it, thanks :)

**Comment by sivchari** [comment]
I'd close this PR once, then we tackle it again after https://github.com/sivchari/govalid/pull/120 and https://github.com/sivchari/govalid/pull/134

**Comment by BlackBuck** [comment]
@sivchari are you planning on re-opening it? You mentioned that you'd merge #120 after this was merged, but you closed this saying you'd merge it after #120 :)
